src/NUnitEngine/nunit.engine.core/Drivers/NUnitFrameworkApi2018.cs

```diff
@@ -0,0 +1,253 @@
+            log.Debug($"Created FrameworkController {_frameworkControllerType?.Name}");
+
+            var controllerAssembly = _frameworkControllerType?.Assembly?.GetName();
+            log.Debug($"Controller assembly is {controllerAssembly}");
+#else
+            _assemblyLoadContext = new TestAssemblyLoadContext(testAssemblyPath);
+
+            var testAssembly = LoadAssembly(testAssemblyPath);
+            _frameworkAssembly = LoadAssembly(_nunitRef);
+
+            _frameworkController = CreateInstance(CONTROLLER_TYPE, testAssembly, idPrefix, settings);
+            if (_frameworkController == null)
+            {
+                log.Error(INVALID_FRAMEWORK_MESSAGE);
+                throw new NUnitEngineException(INVALID_FRAMEWORK_MESSAGE);
+            }
+#endif
+
+            _frameworkControllerType = _frameworkController?.GetType();
+            log.Debug($"Created FrameworkController {_frameworkControllerType?.Name}");
+
+            log.Debug($"Loaded {testAssemblyPath}");
+            return (string)ExecuteMethod(LOAD_METHOD);
+        }
+
+        public int CountTestCases(string filter)
+        {
+            CheckLoadWasCalled();
+            object? count = ExecuteMethod(COUNT_METHOD, filter);
+            return count != null ? (int)count : 0;
+        }
+
+        public string Run(ITestEventListener? listener, string filter)
+        {
+            CheckLoadWasCalled();
+            log.Info("Running {0} - see separate log file", Path.GetFileName(_testAssemblyPath.ShouldNotBeNull()));
+            Action<string>? callback = null;
+            return (string)ExecuteMethod(RUN_METHOD, new[] { typeof(Action<string>), typeof(string) }, callback, filter);
+        }
+
+        public void RunAsync(Action<string> callback, string filter)
+        {
+            CheckLoadWasCalled();
+            log.Info("Running {0} - see separate log file", Path.GetFileName(_testAssemblyPath.ShouldNotBeNull()));
+            ExecuteMethod(RUN_ASYNC_METHOD, new[] { typeof(Action<string>), typeof(string) }, callback, filter);
+        }
+
+        public void StopRun(bool force)
+        {
+            ExecuteMethod(STOP_RUN_METHOD, force);
+        }
+
+        public string Explore(string filter)
+        {
+            CheckLoadWasCalled();
+            log.Info("Exploring {0} - see separate log file", Path.GetFileName(_testAssemblyPath.ShouldNotBeNull()));
+            return (string)ExecuteMethod(EXPLORE_METHOD, filter);
+        }
+
+        private void CheckLoadWasCalled()
+        {
+            if (_frameworkController == null)
+                throw new InvalidOperationException(LOAD_MESSAGE);
+        }
+
+#if NETCOREAPP
+        private object CreateInstance(string typeName, params object?[]? args)
+        {
+            var type = _frameworkAssembly.ShouldNotBeNull().GetType(typeName, throwOnError: true)!;
+            return Activator.CreateInstance(type, args)!;
```

Comment:
This is where the `using (_assemblyLoadContext.ShouldNotBeNull().EnterContextualReflection())`
should be called as dependencies are loaded when calling `GetType` or `CreateInstance`.

Reply:
Not needed in our tests, but it's certainly possible it may be needed in some situation. Maybe try a test that depends on a different version of `TestCentric.Metadata` than the one we use. It's supposed to work as a result of our creating a special context for the tests - equivalent to the NetFramework approach of using a separate AppDomain.